### PR TITLE
[SPARK-7689] Remove TTL-based metadata cleaning in Spark 2.0

### PR DESCRIPTION
**Pull request by JoshRosen**
This PR removes `spark.cleaner.ttl` and the associated TTL-based metadata cleaning code.

Now that we have the `ContextCleaner` and a timer to trigger periodic GCs, I don't think that `spark.cleaner.ttl` is necessary anymore. The TTL-based cleaning isn't enabled by default, isn't included in our end-to-end tests, and has been a source of user confusion when it is misconfigured. If the TTL is set too low, data which is still being used may be evicted / deleted, leading to hard to diagnose bugs.

For all of these reasons, I think that we should remove this functionality in Spark 2.0. Additional benefits of doing this include marginally reduced memory usage, since we no longer need to store timetsamps in hashmaps, and a handful fewer threads.

### CODE REVIEW

**Comment by jerryshao** [comment]
Good to remove it :). It has been obsolete for a long time, and will potentially lead to some unexpected behaviors especially in Streaming (like block not found).

**Comment by rxin** [comment]
LGTM. We should let @tdas take a look at this though.

File: streaming/src/test/scala/org/apache/spark/streaming/StreamingContextSuite.scala [hunk 1]
```diff
@@ -81,6 +81,7 @@ class StreamingContextSuite extends SparkFunSuite with BeforeAndAfter with Timeo
 
   test("from conf with settings") {
     val myConf = SparkContext.updatedConf(new SparkConf(false), master, appName)
+    // TODO(josh): Update these exmaples to use a different configuration.
```
Comment (JoshRosen) [inline streaming/src/test/scala/org/apache/spark/streaming/StreamingContextSuite.scala]: TODO.

File: core/src/main/scala/org/apache/spark/MapOutputTracker.scala [hunk 2]
```diff
@@ -291,17 +289,10 @@ private[spark] class MapOutputTrackerMaster(conf: SparkConf)
...
-   */
-  protected val mapStatuses = new TimeStampedHashMap[Int, Array[MapStatus]]()
-  private val cachedSerializedStatuses = new TimeStampedHashMap[Int, Array[Byte]]()
-
-  // For cleaning up TimeStampedHashMaps
-  private val metadataCleaner =
-    new MetadataCleaner(MetadataCleanerType.MAP_OUTPUT_TRACKER, this.cleanup, conf)
+  // HashMaps for storing mapStatuses and cached serialized statuses in the driver.
+  // Statuses are dropped only by explicit de-registering.
+  protected val mapStatuses = new HashMap[Int, Array[MapStatus]]()
```
Comment (JoshRosen) [inline core/src/main/scala/org/apache/spark/MapOutputTracker.scala]: I think that these need to be ConcurrentHashMaps in order to preserve the old code's thread-safety guarantees.

File: core/src/main/scala/org/apache/spark/shuffle/FileShuffleBlockResolver.scala [hunk 3]
```diff
@@ -63,10 +63,7 @@ private[spark] class FileShuffleBlockResolver(conf: SparkConf)
     val completedMapTasks = new ConcurrentLinkedQueue[Int]()
   }
 
-  private val shuffleStates = new TimeStampedHashMap[ShuffleId, ShuffleState]
-
-  private val metadataCleaner =
-    new MetadataCleaner(MetadataCleanerType.SHUFFLE_BLOCK_MANAGER, this.cleanup, conf)
+  private val shuffleStates = new scala.collection.mutable.HashMap[ShuffleId, ShuffleState]
```
Comment (JoshRosen) [inline core/src/main/scala/org/apache/spark/shuffle/FileShuffleBlockResolver.scala]: I'll also make this into a ConcurrentHashMap.

File: core/src/main/scala/org/apache/spark/SparkContext.scala [hunk 4]
```diff
@@ -295,8 +295,7 @@ class SparkContext(config: SparkConf) extends Logging with ExecutorAllocationCli
   private[spark] val addedJars = HashMap[String, Long]()
 
   // Keeps track of all persisted RDDs
-  private[spark] val persistentRdds = new TimeStampedWeakValueHashMap[Int, RDD[_]]
-  private[spark] def metadataCleaner: MetadataCleaner = _metadataCleaner
+  private[spark] val persistentRdds = new MapMaker().weakValues().makeMap[Int, RDD[_]]().asScala
```
Comment (JoshRosen) [inline core/src/main/scala/org/apache/spark/SparkContext.scala]: I'll add a comment here to clarify that MapMaker returns a ConcurrentHashMap (or will just use an explicit return type, if Guava supports that).

**Comment by JoshRosen** [comment]
Jenkins, retest this please.

**Comment by gatorsmile** [comment]
All the recent test build failed due to the same issue.

**Comment by JoshRosen** [comment]
Jenkins, retest this please.

**Comment by andrewor14** [comment]
+1

File: core/src/main/scala/org/apache/spark/storage/BlockManager.scala [hunk 5]
```diff
@@ -75,7 +77,7 @@ private[spark] class BlockManager(
 
   val diskBlockManager = new DiskBlockManager(this, conf)
 
-  private val blockInfo = new TimeStampedHashMap[BlockId, BlockInfo]
+  private val blockInfo = new ConcurrentHashMap[BlockId, BlockInfo]
```
Comment (tdas) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala]: Isnt it simpler to keep the Scala map interface? Will minimize changes in rest of code in this class.

Comment (JoshRosen) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala]: I was worried about the pitfalls of `.getOrElseUpdate` not being atomic on `ConcurrentHashMaps` that had been wrapped into Scala maps.

File: streaming/src/test/scala/org/apache/spark/streaming/StreamingContextSuite.scala [hunk 6]
```diff
@@ -93,26 +94,27 @@ class StreamingContextSuite extends SparkFunSuite with BeforeAndAfter with Timeo
 
   test("from existing SparkContext with settings") {
     val myConf = SparkContext.updatedConf(new SparkConf(false), master, appName)
-    myConf.set("spark.cleaner.ttl", "10s")
```
Comment (tdas) [inline streaming/src/test/scala/org/apache/spark/streaming/StreamingContextSuite.scala]: These lines are not necessary. So all lines in streaming tests that has cleaner.ttl can be removed.

File: streaming/src/test/scala/org/apache/spark/streaming/StreamingContextSuite.scala [hunk 7]
```diff
@@ -81,9 +81,10 @@ class StreamingContextSuite extends SparkFunSuite with BeforeAndAfter with Timeo
 
   test("from conf with settings") {
     val myConf = SparkContext.updatedConf(new SparkConf(false), master, appName)
-    myConf.set("spark.cleaner.ttl", "10s")
+    // TODO(josh): Update these exmaples to use a different configuration.
+    myConf.set("spark.cleaner.periodicGC.interval", "10s")
```
Comment (tdas) [inline streaming/src/test/scala/org/apache/spark/streaming/StreamingContextSuite.scala]: Better to do this change here already. With a some random spark conf

File: streaming/src/test/scala/org/apache/spark/streaming/StreamingContextSuite.scala (hunk 6 above)
Comment (JoshRosen) [inline streaming/src/test/scala/org/apache/spark/streaming/StreamingContextSuite.scala]: These tests are just testing the configuration persistence system; the actual conf doesn't matter. Want me to replace it with some dummy conf instead?

Comment (tdas) [inline streaming/src/test/scala/org/apache/spark/streaming/StreamingContextSuite.scala]: yeah. dummy conf is less confusing. Thanks for the change.

File: docs/configuration.md [hunk 8]
```diff
@@ -1439,7 +1428,7 @@ Apart from these, the following properties are also available, and may be useful
         <td>
             A comma separated list of ciphers. The specified ciphers must be supported by JVM.
             The reference list of protocols one can find on
-            <a href="https://blogs.oracle.com/java-platform-group/entry/diagnosing_tls_ssl_and_https">this</a>
+            a href="https://blogs.oracle.com/java-platform-group/entry/diagnosing_tls_ssl_and_https">this</a>
```
Comment (tdas) [inline docs/configuration.md]: why is this < removed??

File: core/src/main/scala/org/apache/spark/storage/BlockManager.scala (hunk 5 above)
Comment (tdas) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala]: Good point.

**Comment by tdas** [comment]
LGTM, pending tests.

File: docs/configuration.md (hunk 8 above)
Comment (JoshRosen) [inline docs/configuration.md]: Whoops, thanks for catching that.

**Comment by JoshRosen** [comment]
Jenkins, retest this please.

**Comment by rxin** [comment]
Merging. Thanks.

**Comment by JoshRosen** [comment]
Cross-references to related PRs (to aid other code archaeologists):

- Earlier discussion of deprecating this flag: https://github.com/apache/spark/pull/6220
- The PR which introduced periodic forced driver GCs: https://github.com/apache/spark/pull/10070

**Pull request closed without merging**